### PR TITLE
add skipAnalytics & netrcLogin

### PR DESCRIPTION
**Pull request by KarateCowboy**

### CODE REVIEW

File: src/config.js [hunk 1]
```diff
@@ -63,6 +67,23 @@ function debug () {
...
+    return true
+  } else if (this._skipAnalytics) {
+    return true
+  } else if (this.netrcLogin() === false) {
+    return true
+  }
+  return false
+}
+
+async function netrcLogin () {
```
Comment (ransombriggs) [inline src/config.js]: @KarateCowboy could you move this back to the `cli-engine` and just have the responsibility of the config to check if the `TESTING` env var is set and checking the [user config file](https://github.com/heroku/cli/blob/master/config.go#L8) for which I cannot find the implementation.

File: src/config.js [hunk 2]
```diff
@@ -81,6 +102,7 @@ export function buildConfig (options: ConfigOptions = {}): Config {
     root: path.join(__dirname, '..'),
     platform: os.platform(),
     arch: os.arch(),
+    _skipAnalytics: false,
```
Comment (ransombriggs) [inline src/config.js]: @KarateCowboy continuing comment from above: could this be done like `debug` where it sets the default from the `skipAnalytics()` function?

File: src/config.js [hunk 3]
```diff
@@ -63,6 +67,23 @@ function debug () {
   if (HEROKU_DEBUG) return parseInt(HEROKU_DEBUG)
   return 0
 }
+function skipAnalytics () {
+  if (process.env['TESTING'] && process.env['TESTING'].match(/[\w]+/)) {
```
Comment (KarateCowboy) [inline src/config.js]: Pushed

**Comment by KarateCowboy** [comment]
ready for another round of review

File: package.json [hunk 4]
```diff
@@ -4,7 +4,9 @@
   "version": "1.0.2",
   "author": "Jeff Dickey @dickeyxxx",
   "bugs": "https://github.com/heroku/cli-engine-config/issues",
-  "dependencies": {},
+  "dependencies": {
+    "netrc-parser": "^2.0.1"
```
Comment (ransombriggs) [inline package.json]: @KarateCowboy could you remove this dependency now?

File: src/config.js [hunk 5]
```diff
@@ -63,6 +67,20 @@ function debug () {
...
+    return true
+  } else if (config.userConfig.skipAnalytics) {
+    return true
+  } else {
+    return false
+  }
+}
+
+let loadUserConfig = function (configDir) {
+  const rawdata = fs.readFileSync(`${configDir}/user_config.json`, 'utf8')
```
Comment (ransombriggs) [inline src/config.js]: @KarateCowboy use `path.join` here for cross platform compatibility

Comment (ransombriggs) [inline src/config.js]: @KarateCowboy Can be shortened with [fs.readJsonSync](https://github.com/jprichardson/node-fs-extra/blob/master/docs/readJson-sync.md)

File: package.json (hunk 4 above)
Comment (dickeyxxx) [inline package.json]: Why is this added?

File: src/user-config.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (dickeyxxx) [inline src/user-config.js]: Missing flow directive

File: src/user-config.test.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import { type UserConfig } from './user-config'
```
Comment (dickeyxxx) [inline src/user-config.test.js]: Missing flow directive

File: src/config.js [hunk 8]
```diff
@@ -43,7 +44,10 @@ export type Config = {
   arch: string,             // CPU architecture
   platform: string,         // operating system
   windows: boolean,         // is windows OS
-  _version: '1'             // config schema version
+  _version: '1',             // config schema version
+  userConfig: UserConfig,
```
Comment (ransombriggs) [inline src/config.js]: @KarateCowboy I would not expose `userConfig` here, it adds another redundant way to access `skipAnalytics` and consumers of this should not have to think about whether something is exposed via the user config file or not.

File: src/config.js [hunk 9]
```diff
@@ -43,7 +44,10 @@ export type Config = {
...
+  skipAnalytics: boolean
```
Comment (dickeyxxx) [inline src/config.js]: Would be nice if these were documented like the others

File: src/config.js [hunk 10]
```diff
@@ -63,6 +67,20 @@ function debug () {
...
+  } else if (config.userConfig.skipAnalytics) {
+    return true
+  } else {
+    return false
+  }
+}
+
+let loadUserConfig = function (configDir) {
+  const rawdata = fs.readFileSync(`${configDir}/user_config.json`, 'utf8')
+  return JSON.parse(rawdata)
```
Comment (dickeyxxx) [inline src/config.js]: What happens if the file doesn't exist?

File: src/config.js (hunk 8 above)
Comment (dickeyxxx) [inline src/config.js]: +1

File: package.json (hunk 4 above)
Comment (KarateCowboy) [inline package.json]: Originallly had the netrc stuff in here, but yanked it back out. This dependency is now removed.

File: src/config.js (hunk 10 above)
Comment (KarateCowboy) [inline src/config.js]: I now have it trapped if it's a file-not-found error. Otherwise, rethrow.

File: src/user-config.test.js (hunk 7 above)
Comment (KarateCowboy) [inline src/user-config.test.js]: fixed

File: src/config.js (hunk 9 above)
Comment (KarateCowboy) [inline src/config.js]: fixed

File: src/user-config.js (hunk 6 above)
Comment (KarateCowboy) [inline src/user-config.js]: fixed

File: src/config.js (hunk 5 above)
Comment (KarateCowboy) [inline src/config.js]: fixed

File: src/config.js (hunk 8 above)
Comment (KarateCowboy) [inline src/config.js]: fixed

File: package.json (hunk 4 above)
Comment (KarateCowboy) [inline package.json]: fixed

File: src/config.js [hunk 11]
```diff
@@ -63,7 +65,29 @@ function debug () {
...
+  }
+  return false
+}
+
+let loadUserConfig = function (configDir) {
+  try {
+    const config: UserConfig = fs.readJsonSync(path.join(configDir, 'user_config.json'), 'utf8')
+    return config
+  } catch (e) {
+    if (e.message.match(/no such file/)) {
```
Comment (ransombriggs) [inline src/config.js]: @KarateCowboy could you check `e.code === 'ENOENT'` instead?

File: src/config.js [hunk 12]
```diff
@@ -63,7 +65,29 @@ function debug () {
   if (HEROKU_DEBUG) return parseInt(HEROKU_DEBUG)
   return 0
 }
+function skipAnalytics (userConfig: UserConfig) {
+  if (userConfig.skipAnalytics) {
+    return true
+  } else if (process.env['TESTING'] && process.env['TESTING'] === '1') {
```
Comment (ransombriggs) [inline src/config.js]: @KarateCowboy pretty sure that the `process.env['TESTING']` is redundant here

File: package.json [hunk 13]
```diff
@@ -4,7 +4,9 @@
   "version": "1.0.2",
   "author": "Jeff Dickey @dickeyxxx",
   "bugs": "https://github.com/heroku/cli-engine-config/issues",
-  "dependencies": {},
+  "dependencies": {
+    "cli-engine-config": "^1.0.3"
```
Comment (ransombriggs) [inline package.json]: @KarateCowboy what is this for?

Comment (KarateCowboy) [inline package.json]: Not sure. I thought it was something yarn did automagically. I'll remove it if it's not necessary.

File: src/config.js (hunk 12 above)
Comment (KarateCowboy) [inline src/config.js]: fixed

File: src/config.js (hunk 11 above)
Comment (KarateCowboy) [inline src/config.js]: fixed

File: .gitignore [hunk 14]
```diff
@@ -1,3 +1,6 @@
 lib
 node_modules
 coverage
+*swp
+.idea
+yarn-error.log
```
Comment (dickeyxxx) [inline .gitignore]: maybe *.log instead?

File: .gitignore [hunk 15]
```diff
@@ -1,3 +1,6 @@
 lib
 node_modules
 coverage
+*swp
+.idea
```
Comment (dickeyxxx) [inline .gitignore]: these should go in your global gitignore since they're not project specific

File: package.json [hunk 16]
```diff
@@ -1,10 +1,9 @@
 {
   "name": "cli-engine-config",
   "description": "base cli-engine config objects and interfaces",
-  "version": "1.0.2",
+  "version": "1.1.0",
```
Comment (dickeyxxx) [inline package.json]: this should be bumped with `np` when we push everything not in a PR

File: src/config.js [hunk 17]
```diff
@@ -43,7 +44,8 @@ export type Config = {
   arch: string,             // CPU architecture
   platform: string,         // operating system
   windows: boolean,         // is windows OS
-  _version: '1'             // config schema version
+  _version: '1',             // config schema version
```
Comment (dickeyxxx) [inline src/config.js]: whitespace

File: src/config.js [hunk 18]
```diff
@@ -63,7 +65,29 @@ function debug () {
...
+  } catch (e) {
+    if (e.code === 'ENOENT') {
+      return { 'skipAnalytics': false }
+    } else {
+      throw e
+    }
+  }
+}
 
+let userConfig : UserConfig
```
Comment (dickeyxxx) [inline src/config.js]: while for the CLI it's fine to make this a global variable, if we reused this library for chatops we would have to localize it. Better to define the variable inside `buildConfig` I think

Comment (ransombriggs) [inline src/config.js]: @KarateCowboy this should be a local variable in buildConfig

File: src/config.js [hunk 19]
```diff
@@ -82,7 +106,8 @@ export function buildConfig (options: ConfigOptions = {}): Config {
     platform: os.platform(),
     arch: os.arch(),
     bin: cli.bin || 'cli-engine',
-    defaultCommand: cli.defaultCommand || 'help'
+    defaultCommand: cli.defaultCommand || 'help',
+    skipAnalytics: false
```
Comment (ransombriggs) [inline src/config.js]: @KarateCowboy this should be `skipAnalytics: skipAnalytics(userConfig)`

File: src/config.js [hunk 20]
```diff
@@ -91,6 +116,8 @@ export function buildConfig (options: ConfigOptions = {}): Config {
   let defaultCacheDir = process.platform === 'darwin' ? path.join(config.home, 'Library', 'Caches') : null
   config.cacheDir = config.cacheDir || dir(config, 'cache', defaultCacheDir)
   config._version = '1'
+  userConfig = loadUserConfig(config.configDir)
+  config.skipAnalytics = skipAnalytics(userConfig)
```
Comment (ransombriggs) [inline src/config.js]: @KarateCowboy move the userConfig above the defaults and make a local variable and remove the assignment.  The reason is that if someone passes in `skipAnalytics` to `buildConfig` then this will overwrite what they pass rather than being the default

File: package.json (hunk 16 above)
Comment (KarateCowboy) [inline package.json]: kk

File: src/config.js (hunk 19 above)
Comment (KarateCowboy) [inline src/config.js]: I was unable to do this because skipAnalytics requires the userConfig which requires a bunch of stuff created after the config is initialized. Nevertheless, I was able to work in giving precedence to `skipAnalytics` when passed as a 'constructor' argument

File: .gitignore (hunk 14 above)
Comment (KarateCowboy) [inline .gitignore]: kk

File: .gitignore (hunk 15 above)
Comment (KarateCowboy) [inline .gitignore]: kk

File: src/config.js (hunk 17 above)
Comment (KarateCowboy) [inline src/config.js]: kk

File: src/config.js (hunk 18 above)
Comment (KarateCowboy) [inline src/config.js]: kk

File: src/config.js [hunk 21]
```diff
@@ -63,8 +65,30 @@ function debug () {
...
+    return true
+  } else if (process.env['TESTING'] === '1') {
+    return true
+  }
+  return false
+}
+
+let loadUserConfig = function (configDir) {
+  try {
+    const config: UserConfig = fs.readJsonSync(path.join(configDir, 'user_config.json'), 'utf8')
```
Comment (dickeyxxx) [inline src/config.js]: Is this what we called the file before? config.json seems better to me

File: src/config.js [hunk 22]
```diff
@@ -91,6 +116,10 @@ export function buildConfig (options: ConfigOptions = {}): Config {
   let defaultCacheDir = process.platform === 'darwin' ? path.join(config.home, 'Library', 'Caches') : null
   config.cacheDir = config.cacheDir || dir(config, 'cache', defaultCacheDir)
   config._version = '1'
+  userConfig = loadUserConfig(config.configDir)
+  if (config.skipAnalytics === undefined) {
```
Comment (dickeyxxx) [inline src/config.js]: if it is undefined then skipAnalytics would always return false right?

File: src/config.js (hunk 21 above)
Comment (ransombriggs) [inline src/config.js]: @KarateCowboy It is not what we called the file, it was `config.json` https://github.com/heroku/cli/blob/master/config.go#L19

File: src/config.js (hunk 22 above)
Comment (KarateCowboy) [inline src/config.js]: it could still read 'true' from either the user config or the TESTING env var

File: src/config.js (hunk 21 above)
Comment (KarateCowboy) [inline src/config.js]: changed
